README.md
# api-session

![Build Status](https://github.com/janis-commerce/api-session/workflows/Build%20Status/badge.svg)
[![Coverage Status](https://coveralls.io/repos/github/janis-commerce/api-session/badge.svg?branch=master)](https://coveralls.io/github/janis-commerce/api-session?branch=master)
[![npm version](https://badge.fury.io/js/%40janiscommerce%2Fapi-session.svg)](https://www.npmjs.com/package/@janiscommerce/api-session)


A session manager for APIs

## 📦 Installation
```sh
npm install @janiscommerce/api-session
```

## :gear: API
The package exports two classes ApiSession and ApiSessionError.

### `constructor(authorizationData, client)`

Creates an APISession with the `authorizationData` provided or the `client` for direct injection.

#### Parameters

- `authorizationData` is an **optional** _object_ with the following (also optional) properties: { userId, clientId, clientCode, profileId, permissions, hasAccessToAllLocations, locations, warehousesIds }
- `client` is an **optional** _object_ for client injection without performing any DB gets

### `validateLocation(locationId)`

Validate if the location given is valid for the session. It validates against the `locations` and the `hasAccessToAllLocations` boolean.
Returns *Boolean*.

### `validateWarehouse(warehouseId)`. _Since 3.3.0_

Validate if the warehouse given is valid for the session. It validates against the `warehousesIds` and the `hasAccessToAllLocations` boolean.
Returns *Boolean*.

### APISession getters

ApiSession has the following getters:

* userId {string} The ID of the user or undefined in case there is no user
* userIsDev {boolean} If user is dev
* serviceName {string} The name of the service or undefined in case there is no service
* isService {boolean} If session is associated to a service
* clientId {string} The ID of the client or undefined in case there is no client
* clientCode {string} The code of the client or undefined in case there is no client
* currency {string|undefined} The currency of the client or undefined in case there is no client nor currency related. _Since 3.4.0_
* currencyDisplay {string} The currency display of the client or default value in case there is no client. Possible values: `code`, `symbol`. Default: `symbol`. _Since 3.4.0_
* profileId {string} The ID of the profile or undefined in case there is no profile
* hasAccessToAllLocations {boolean} If has access to all locations
* locations {array<string>} The List of Locations to which the user has permissions
* warehousesIds {array<string>} The List of Warehouses to which the user has permissions. _Since 3.3.0_
* permissions {array} The permission keys or undefined in case there are no permissions associated
* *async* client {object} Resolves to the client object with the `getInstance()` method injected. The properties depend on your client internal structure. The client is injected with a `getInstance()` method to propagate the session to other instances.

## Model Client
The package uses the Client Model in our service for getting the clients. For more information see [@janiscommerce/model](https://www.npmjs.com/package/@janiscommerce/model)

## Usage
```js
const { ApiSession, ApiSessionError } = require('@janiscommerce/api-session');
```

## Examples
```js
const { ApiSession } = require('@janiscommerce/api-session');

const SomeModel = require('../models/some-model');

const session = new ApiSession({
	userId: 1,
	userIsDev: false,
	clientId: 2,
	clientCode: 'janis',
	profileId: 5,
	permissions: [
		'catalog:product:read',
		'catalog:product:write'
	],
	locations: ['location-1'],
	hasAccessToAllLocations: false
});

console.log(`Session created for user ${session.userId} on client ${session.clientCode}.`);

const sessionInjectedModel = session.getSessionInstance(SomeModel, 'some-param', 'some-other-param');

console.log(`Session is propagated for user ${sessionInjectedModel.session.userId} on client ${sessionInjectedModel.session.clientCode}.`);

const client = await sessionInjectedModel.session.client;

console.log(client);
// Outputs your client object

const hasAccess = session.validateLocation('location-1');

console.log(`Session has access to location 1: ${hasAccess}`);
// Outputs 'Session has access to location 1: true'
```

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [3.4.0] - 2023-03-20
### Added
- Getters for session fields `currency` and `currencyDisplay`

## [3.3.1] - 2022-03-15
### Removed
- Field hasAccessToAllWarehouses from authentication

### Changed
- Now `validateWarehouse()` will return **true** when `hasAccessToAllLocations` is **true**

## [3.3.0] - 2022-03-11
### Added
- `warehousesIds` and `hasAccessToAllWarehouses` getters
- `validateWarehouse` method

## [3.2.0] - 2021-11-18
### Added
- Getter and setter

## [3.1.1] - 2021-01-27
### Added
- Typings build from JSDoc

## [3.1.0] - 2020-12-17
### Added
- Extra parameters for constructor when 'sessionate' a Class with `getSessionInstance()`

## [3.0.0] - 2020-08-27
### Added
- APISession constructor new parameter `client` object
- GitHub Actions

### Changed
- Using service model for getting clients

### Removed
- Included model and model dependency
- Settings usage

## [2.0.0] - 2020-06-11
### Added
- `locations` and `hasAccessToAllLocations` getters
- `validateLocation` method

### Removed
- `stores` and `hasAccessToAllStores` getters
- `validateStore` method

## [1.4.0] - 2020-05-19
### Removed
- `package-lock.json` file

## [1.3.1] - 2020-03-25
### Added
- `userIsDev`, `serviceName`, `isService` getters

## [1.3.0] - 2020-01-21
### Added
- `stores` and `hasAccessToAllStores` getters
- `validateStore`

## [1.2.1] - 2019-10-16
### Fixed
- README typo fixed

## [1.2.0] - 2019-10-01
### Changed
- Model version updated

## [1.1.0] - 2019-10-01
### Added
- getSessionInstance method to session instance

## [1.0.0] - 2019-09-27
### Added
- Session management
- Client fetching on demand
- Client injection to propagate the session


tests/api-session.js
'use strict';

const assert = require('assert');
const sinon = require('sinon');

const mockRequire = require('mock-require');

const { ApiSession, ApiSessionError } = require('../lib');
const Client = require('./../lib/client');

describe('Api Session', () => {

	afterEach(() => {
		sinon.restore();
		Client._instance = null; // eslint-disable-line
		Client._clientsCache = null; // eslint-disable-line
	});

	const ClientModel = class ClientModel {

		async getBy() {
			// nothing to do here, just for stub
		}
	};

	const loadModelClient = clients => {

		mockRequire(Client.getRelativePath(), ClientModel);

		if(clients) {
			sinon.stub(ClientModel.prototype, 'getBy')
				.resolves(clients);
		}
	};

	context('No authentication data', () => {

		let session;
		beforeEach(() => {
			session = new ApiSession();
		});

		describe('Getters', () => {
			it('Should return undefined for userId', () => {
				assert.strictEqual(session.userId, undefined);
			});

			it('Should return undefined for clientId', () => {
				assert.strictEqual(session.clientId, undefined);
			});

			it('Should return undefined for clientCode', () => {
				assert.strictEqual(session.clientCode, undefined);
			});

			it('Should return undefined for currency', () => {
				assert.strictEqual(session.currency, undefined);
			});

			it('Should return undefined for currencyDisplay', () => {
				assert.strictEqual(session.currencyDisplay, undefined);
			});

			it('Should return undefined for profileId', () => {
				assert.strictEqual(session.profileId, undefined);
			});

			it('Should return undefined for permissions', () => {
				assert.strictEqual(session.permissions, undefined);
			});

			it('Should return undefined for client', async () => {
				assert.strictEqual(await session.client, undefined);
			});

			it('Should return undefined for locations', async () => {
				assert.strictEqual(session.locations, undefined);
			});

			it('Should return undefined for hasAccessToAllLocations', async () => {
				assert.strictEqual(session.hasAccessToAllLocations, undefined);
			});

			it('Should return false for userIsDev', () => {
				assert.strictEqual(session.userIsDev, false);
			});

			it('Should return undefined for serviceName', () => {
				assert.strictEqual(session.serviceName, undefined);
			});

			it('Should return false for isService', () => {
				assert.strictEqual(session.isService, false);
			});

			it('Should return undefined for warehousesIds', async () => {
				assert.strictEqual(session.warehousesIds, undefined);
			});
		});

		describe('Validate Locations', () => {
			it('Should return false', () => {
				assert.strictEqual(session.validateLocation('622b7077e1a195d5c2dd9627'), false);
			});
		});

		describe('Validate Warehouses', () => {
			it('Should return false', () => {
				assert.strictEqual(session.validateWarehouse('622b7082681fce264b21aefd'), false);
			});
		});
	});

	context('User related authentication data', () => {

		const locationId1 = '622b709ad74dba38ce55e961';
		const locationId2 = '622b70a27bdb6f41005cedda';
		const locationId3 = '622b71aba86eca42cc08b526';
		const locationId4 = '622b71b3c95364ef129b7649';

		const warehouseId1 = '622b70bb68c06073782f2a30';
		const warehouseId2 = '622b70bf2926e982ea3d1cbd';
		const warehouseId3 = '622b7200dbd397165b2fd4f0';
		const warehouseId4 = '622b7204c7b2e81e7121dd41';

		let session;

		beforeEach(() => {
			session = new ApiSession({
				serviceName: 'some-service',
				userId: 'some-user-id',
				userIsDev: true,
				clientId: 'some-client-id',
				clientCode: 'some-client-code',
				currency: 'EUR',
				currencyDisplay: 'code',
				profileId: 'some-profile-id',
				permissions: ['service:namespace:method1', 'service:namespace:method2'],
				locations: [locationId1, locationId2],
				hasAccessToAllLocations: false,
				warehousesIds: [warehouseId1, warehouseId2]
			});
		});

		describe('Getters', () => {
			it('Should return the correct userId', () => {
				assert.strictEqual(session.userId, 'some-user-id');
			});

			it('Should return the correct clientId', () => {
				assert.strictEqual(session.clientId, 'some-client-id');
			});

			it('Should return the correct clientCode', () => {
				assert.strictEqual(session.clientCode, 'some-client-code');
			});

			it('Should return the correct currency', () => {
				assert.strictEqual(session.currency, 'EUR');
			});

			it('Should return the correct currencyDisplay', () => {
				assert.strictEqual(session.currencyDisplay, 'code');
			});

			it('Should return the correct profileId', () => {
				assert.strictEqual(session.profileId, 'some-profile-id');
			});

			it('Should return the correct userIsDev', () => {
				assert.strictEqual(session.userIsDev, true);
			});

			it('Should return the correct serviceName', () => {
				assert.strictEqual(session.serviceName, 'some-service');
			});

			it('Should return the correct isService', () => {
				assert.strictEqual(session.isService, true);
			});

			it('Should return the correct permissions', () => {
				assert.deepStrictEqual(session.permissions, ['service:namespace:method1', 'service:namespace:method2']);
			});

			it('Should return the correct locations', () => {
				assert.deepStrictEqual(session.locations, ['622b709ad74dba38ce55e961', '622b70a27bdb6f41005cedda']);
			});

			it('Should return the correct hasAccessToAllLocations', () => {
				assert.strictEqual(session.hasAccessToAllLocations, false);
			});

			it('Should return the correct warehousesIds', () => {
				assert.deepStrictEqual(session.warehousesIds, ['622b70bb68c06073782f2a30', '622b70bf2926e982ea3d1cbd']);
			});
		});

		describe('Client getter', () => {

			it('Should throw if model client not found', async () => {

				await assert.rejects(() => session.client, {
					name: 'ApiSessionError',
					code: ApiSessionError.codes.INVALID_MODEL_CLIENT
				});
			});

			it('Should throw if model client rejects', async () => {

				loadModelClient();

				sinon.stub(ClientModel.prototype, 'getBy')
					.rejects('Some model error');

				await assert.rejects(() => session.client, {
					name: 'ApiSessionError',
					code: ApiSessionError.codes.INTERNAL_ERROR
				});

				sinon.assert.calledWithExactly(ClientModel.prototype.getBy, 'code', 'some-client-code', { limit: 1 });
			});

			it('Should throw if model client can\'t found the client', async () => {

				loadModelClient([]);

				await assert.rejects(() => session.client, {
					name: 'ApiSessionError',
					code: ApiSessionError.codes.CLIENT_NOT_FOUND
				});

				sinon.assert.calledWithExactly(ClientModel.prototype.getBy, 'code', 'some-client-code', { limit: 1 });
			});

			it('Should return the injected client with a working getInstance', async () => {

				const baseClient = {
					id: 'the-client-id-123',
					code: 'some-client-code',
					otherData: 876
				};

				loadModelClient([baseClient]);

				const client = await session.client;

				sinon.assert.match(client, {
					...baseClient,
					getInstance: sinon.match.func
				});

				sinon.assert.calledWithExactly(ClientModel.prototype.getBy, 'code', 'some-client-code', { limit: 1 });

				class Test {}

				const testInstance = client.getInstance(Test);

				assert.deepStrictEqual(testInstance.session, session);
			});

			it('Should return client without getting if APISession instanced with client object', async () => {

				loadModelClient();

				sinon.stub(ClientModel.prototype, 'getBy');

				const offlineClient = {
					id: 'client-id-3',
					code: 'offline-client'
				};

				const offlineSession = new ApiSession({}, offlineClient);

				const client = await offlineSession.client;

				sinon.assert.match(client, {
					...offlineClient,
					getInstance: sinon.match.func
				});

				sinon.assert.notCalled(ClientModel.prototype.getBy);

				class Test {}

				const testInstance = client.getInstance(Test);

				assert.deepStrictEqual(testInstance.session, offlineSession);
			});

		});

		describe('getSessionInstance', () => {

			it('Should return an instance with the session injected', () => {

				class Test {}

				const testInstance = session.getSessionInstance(Test);

				assert.deepStrictEqual(testInstance.session, session);
			});
		});

		describe('Client cache', () => {

			it('Should fetch the client from de DB only once by default', async () => {

				sinon.useFakeTimers(Date.now() + (15 * 60 * 1000)); // Both requests will be issued "in 15 minutes"

				const baseClient = {
					id: 'the-client-id-123',
					code: 'some-client-code',
					otherData: 876
				};

				loadModelClient([baseClient]);

				sinon.assert.match(await session.client, {
					...baseClient,
					getInstance: sinon.match.func
				});

				sinon.assert.match(await session.client, {
					...baseClient,
					getInstance: sinon.match.func
				});

				sinon.assert.calledOnceWithExactly(ClientModel.prototype.getBy, 'code', 'some-client-code', { limit: 1 });
			});

			it('Should fetch the client from de DB again after 10 minutes by default', async () => {

				const now = Date.now();

				sinon.stub(Date, 'now')
					.onCall(0)
					.returns(now) // First requests will be issued "now"
					.onCall(1)
					.returns(now + (15 * 60 * 1000)) // First requests will be issued "in 15 minutes"
					.onCall(2)
					.returns(now + (20 * 60 * 1000)); // Second requests will be issued "in 20 minutes"

				const baseClient = {
					id: 'the-client-id-123',
					code: 'some-client-code',
					otherData: 876
				};

				loadModelClient([baseClient]);

				sinon.assert.match(await session.client, {
					...baseClient,
					getInstance: sinon.match.func
				});

				sinon.assert.match(await session.client, {
					...baseClient,
					getInstance: sinon.match.func
				});

				sinon.assert.match(await session.client, {
					...baseClient,
					getInstance: sinon.match.func
				});

				sinon.assert.calledTwice(ClientModel.prototype.getBy);
				sinon.assert.calledWithExactly(ClientModel.prototype.getBy, 'code', 'some-client-code', { limit: 1 });
			});

			it('Should fetch the client from de DB again if useClientCache flag is set as false', async () => {

				sinon.useFakeTimers(Date.now() + (15 * 60 * 1000)); // Both requests will be issued "in 15 minutes"

				const baseClient = {
					id: 'the-client-id-123',
					code: 'some-client-code',
					otherData: 876
				};

				loadModelClient([baseClient]);

				// Turn cache off
				session.useClientCache = false;

				sinon.assert.match(await session.client, {
					...baseClient,
					getInstance: sinon.match.func
				});

				sinon.assert.match(await session.client, {
					...baseClient,
					getInstance: sinon.match.func
				});

				sinon.assert.match(await session.client, {
					...baseClient,
					getInstance: sinon.match.func
				});

				sinon.assert.calledThrice(ClientModel.prototype.getBy);
				sinon.assert.calledWithExactly(ClientModel.prototype.getBy, 'code', 'some-client-code', { limit: 1 });
			});

		});

		describe('Validate Location', () => {

			it('Should return false when session has not access to all location and no locationId is passed', () => {
				assert.strictEqual(session.validateLocation(), false);
			});

			it('Should return false when session has not locations field and hasAccessToAllLocations as false', () => {

				const invalidSession = new ApiSession({
					userId: 'some-user-id',
					clientId: 'some-client-id',
					clientCode: 'some-client-code',
					profileId: 'some-profile-id',
					hasAccessToAllLocations: false,
					permissions: ['service:namespace:method1', 'service:namespace:method2']
				});

				assert.strictEqual(invalidSession.validateLocation(locationId1), false);
			});

			it('Should return false when session has not valid locations field and hasAccessToAllLocations as false', () => {

				const invalidSession = new ApiSession({
					userId: 'some-user-id',
					clientId: 'some-client-id',
					clientCode: 'some-client-code',
					profileId: 'some-profile-id',
					permissions: ['service:namespace:method1', 'service:namespace:method2'],
					hasAccessToAllLocations: false,
					locations: { 1: locationId1, 2: locationId2 }
				});

				assert.strictEqual(invalidSession.validateLocation(locationId1), false);
			});

			it('Should return false when session has an empty array of locations and hasAccessToAllLocations as false', () => {

				const invalidSession = new ApiSession({
					userId: 'some-user-id',
					clientId: 'some-client-id',
					clientCode: 'some-client-code',
					profileId: 'some-profile-id',
					permissions: ['service:namespace:method1', 'service:namespace:method2'],
					hasAccessToAllLocations: false,
					locations: []
				});

				assert.strictEqual(invalidSession.validateLocation(locationId1), false);
			});

			it('Should return true when session has access to that location', () => {
				assert.strictEqual(session.validateLocation(locationId1), true);
				assert.strictEqual(session.validateLocation(locationId2), true);
			});

			it('Should return false when session has no access to that location', () => {
				assert.strictEqual(session.validateLocation(locationId3), false);
				assert.strictEqual(session.validateLocation(locationId4), false);
			});

			it('Should return true when session has access to all locations', () => {

				const invalidSession = new ApiSession({
					userId: 'some-user-id',
					clientId: 'some-client-id',
					clientCode: 'some-client-code',
					profileId: 'some-profile-id',
					permissions: ['service:namespace:method1', 'service:namespace:method2'],
					hasAccessToAllLocations: true
				});

				assert.strictEqual(invalidSession.validateLocation(locationId4), true);
			});
		});

		describe('Validate Warehouse', () => {

			it('Should return false when session has not access to all locations and no warehouseId is passed', () => {
				assert.strictEqual(session.validateWarehouse(), false);
			});

			it('Should return false when session has not warehouses field and hasAccessToAllLocations as false', () => {

				const invalidSession = new ApiSession({
					userId: 'some-user-id',
					clientId: 'some-client-id',
					clientCode: 'some-client-code',
					profileId: 'some-profile-id',
					permissions: ['service:namespace:method1', 'service:namespace:method2'],
					hasAccessToAllLocations: false
				});

				assert.strictEqual(invalidSession.validateWarehouse(warehouseId1), false);
			});

			it('Should return false when session has not valid warehousesIds field and hasAccessToAllLocations as false', () => {

				const invalidSession = new ApiSession({
					userId: 'some-user-id',
					clientId: 'some-client-id',
					clientCode: 'some-client-code',
					profileId: 'some-profile-id',
					permissions: ['service:namespace:method1', 'service:namespace:method2'],
					hasAccessToAllLocations: false,
					warehousesIds: { 1: warehouseId1, 2: warehouseId2 }
				});

				assert.strictEqual(invalidSession.validateWarehouse(warehouseId1), false);
			});

			it('Should return false when session has an empty array of warehousesIds and hasAccessToAllLocations as false', () => {

				const invalidSession = new ApiSession({
					userId: 'some-user-id',
					clientId: 'some-client-id',
					clientCode: 'some-client-code',
					profileId: 'some-profile-id',
					permissions: ['service:namespace:method1', 'service:namespace:method2'],
					hasAccessToAllLocations: false,
					warehousesIds: []
				});

				assert.strictEqual(invalidSession.validateWarehouse(warehouseId1), false);
			});

			it('Should return true when session has access to that warehouse', () => {
				assert.strictEqual(session.validateWarehouse(warehouseId1), true);
				assert.strictEqual(session.validateWarehouse(warehouseId2), true);
			});

			it('Should return false when session has no access to that warehouse', () => {
				assert.strictEqual(session.validateWarehouse(warehouseId3), false);
				assert.strictEqual(session.validateWarehouse(warehouseId4), false);
			});

			it('Should return true when session has access to all locations', () => {

				const invalidSession = new ApiSession({
					userId: 'some-user-id',
					clientId: 'some-client-id',
					clientCode: 'some-client-code',
					profileId: 'some-profile-id',
					permissions: ['service:namespace:method1', 'service:namespace:method2'],
					hasAccessToAllLocations: true
				});

				assert.strictEqual(invalidSession.validateWarehouse(warehouseId4), true);
			});
		});
	});
});


tests/model-client.js
'use strict';

const assert = require('assert');
const sinon = require('sinon');

const Model = require('@janiscommerce/model');
const Settings = require('@janiscommerce/settings');

const ModelClient = require('./../lib/model-client');

describe('Client Model', () => {

	const modelClient = new ModelClient();

	afterEach(() => {
		sinon.restore();
	});

	describe('Database key getter', () => {

		beforeEach(() => {
			sinon.stub(Settings, 'get');
		});

		it('Should return the default key if settings are not defined', () => {
			assert.strictEqual(modelClient.databaseKey, 'core');
		});

		it('Should return the default key if databaseKey setting is not defined', () => {

			Settings.get.returns({ foo: 'bar' });

			assert.strictEqual(modelClient.databaseKey, 'core');
		});

		it('Should return the key from the settings if it\'s present', () => {

			Settings.get.returns({ databaseKey: 'custom-key' });

			assert.strictEqual(modelClient.databaseKey, 'custom-key');
		});
	});

	describe('Table static getter', () => {

		beforeEach(() => {
			sinon.stub(Settings, 'get');
		});

		it('Should return the default table if settings are not defined', () => {
			assert.strictEqual(ModelClient.table, 'clients');
		});

		it('Should return the default table if table setting is not defined', () => {

			Settings.get.returns({ foo: 'bar' });

			assert.strictEqual(ModelClient.table, 'clients');
		});

		it('Should return the table from the settings if it\'s present', () => {

			Settings.get.returns({ table: 'custom-table' });

			assert.strictEqual(ModelClient.table, 'custom-table');
		});
	});

	describe('Identifier field static getter', () => {

		beforeEach(() => {
			sinon.stub(Settings, 'get');
		});

		it('Should return the default field if settings are not defined', () => {
			assert.strictEqual(ModelClient.identifierField, 'code');
		});

		it('Should return the default field if identifierField setting is not defined', () => {

			Settings.get.returns({ foo: 'bar' });

			assert.strictEqual(ModelClient.identifierField, 'code');
		});

		it('Should return the field from the settings if it\'s present', () => {

			Settings.get.returns({ identifierField: 'custom-field' });

			assert.strictEqual(ModelClient.identifierField, 'custom-field');
		});
	});

	describe('Fields static getter', () => {

		it('Should return the model available fields', () => {
			assert.deepStrictEqual(ModelClient.fields, {
				code: true
			});
		});
	});

	describe('getByField', () => {

		beforeEach(() => {
			sinon.stub(Model.prototype, 'get');
		});

		it('Should not fetch the client if field is not passed', async () => {

			const client = await modelClient.getByField('', 'fizzmod');

			assert.strictEqual(client, undefined);
		});

		it('Should not fetch the client if value is not passed', async () => {

			const client = await modelClient.getByField('code', '');

			assert.strictEqual(client, undefined);
		});

		it('Should reject if model.get() fails', async () => {

			Model.prototype.get.rejects('Some internal error');

			await assert.rejects(() => modelClient.getByField('code', 'fizzmod'));

			sinon.assert.calledOnce(Model.prototype.get);
			sinon.assert.calledWithExactly(Model.prototype.get, {
				filters: {
					code: 'fizzmod'
				},
				limit: 1
			});
		});

		it('Should call model.get() with the proper arguments and return the client', async () => {

			const clientMock = { foo: 'bar' };

			Model.prototype.get.resolves({ ...clientMock });

			const client = await modelClient.getByField('code', 'fizzmod');

			assert.deepStrictEqual(client, clientMock);

			sinon.assert.calledOnce(Model.prototype.get);
			sinon.assert.calledWithExactly(Model.prototype.get, {
				filters: {
					code: 'fizzmod'
				},
				limit: 1
			});
		});
	});

});


tests/api-session-error.js
'use strict';

const assert = require('assert');

const { ApiSessionError } = require('../lib');

describe('Api Session Error', () => {

	it('Should accept a message error and a code', () => {
		const error = new ApiSessionError('Some error', ApiSessionError.codes.INTERNAL_ERROR);

		assert.strictEqual(error.message, 'Some error');
		assert.strictEqual(error.code, ApiSessionError.codes.INTERNAL_ERROR);
		assert.strictEqual(error.name, 'ApiSessionError');
	});

	it('Should accept an error instance and a code', () => {

		const previousError = new Error('Some error');

		const error = new ApiSessionError(previousError, ApiSessionError.codes.INTERNAL_ERROR);

		assert.strictEqual(error.message, 'Some error');
		assert.strictEqual(error.code, ApiSessionError.codes.INTERNAL_ERROR);
		assert.strictEqual(error.name, 'ApiSessionError');
		assert.strictEqual(error.previousError, previousError);
	});
});


lib/api-session.js
'use strict';

const ClientHandler = require('./client');

/**
 * @typedef AuthenticationData
 * @property {string | undefined} userId
 * @property {string | undefined} clientId
 * @property {string | undefined} clientCode
 * @property {string | undefined} profileId
 * @property {string | undefined} serviceName
 * @property {Array<string> | undefined} permissions
 * @property {Array<string> | undefined} locations
 */

module.exports = class ApiSession {

	/**
	 * The class constructor
	 *
	 * @param {AuthenticationData} authenticationData
	 * @param {import('./client').Client} client
	 */
	constructor(authenticationData, client) {
		/** @private */
		this.authenticationData = authenticationData || {};
		/** @private */
		this._client = client;
		/** @private */
		this._useClientCache = true;
	}

	get useClientCache() {
		return this._useClientCache;
	}

	set useClientCache(useClientCache) {
		this._useClientCache = useClientCache;
	}

	/**
	 * Get the user ID associated to the session
	 *
	 * @return {string|undefined} The ID of the user or undefined in case there is no user
	 */
	get userId() {
		return this.authenticationData.userId;
	}

	/**
	 * Get if user is dev associated to the session
	 *
	 * @return {boolean}
	 */
	get userIsDev() {
		return !!this.authenticationData.userIsDev;
	}

	/**
	 * Get the client ID associated to the session
	 *
	 * @return {string|undefined} The ID of the client or undefined in case there is no client
	 */
	get clientId() {
		return this.authenticationData.clientId;
	}

	/**
	 * Get the client code associated to the session
	 *
	 * @return {string|undefined} The code of the client or undefined in case there is no client
	 */
	get clientCode() {
		return this.authenticationData.clientCode;
	}

	/**
	 * Get the client currency code associated to the session
	 *
	 * @return {string|undefined} The currency code of the client or undefined in case there is no client nor currency
	 */
	get currency() {
		return this.authenticationData.currency;
	}

	/**
	 * Get the client currencyDisplay code associated to the session
	 *
	 * @return {string|undefined} The currencyDisplay code of the client or undefined in case there is no client nor currencyDisplay
	 */
	get currencyDisplay() {
		return this.authenticationData.currencyDisplay;
	}

	/**
	 * Get the profile ID associated to the session
	 *
	 * @return {string|undefined} The ID of the profile or undefined in case there is no profile
	 */
	get profileId() {
		return this.authenticationData.profileId;
	}

	/**
	 * Get the service name associated to the session
	 *
	 * @return {string|undefined} The service name or undefined in case there is no service
	 */
	get serviceName() {
		return this.authenticationData.serviceName;
	}

	/**
	 * Get if the session is associated to a service
	 *
	 * @return {boolean}
	 */
	get isService() {
		return !!this.authenticationData.serviceName;
	}

	/**
	 * Get the permissions array associated to the session
	 *
	 * @return {Array<string>|undefined} The permission keys or undefined in case there are no permissions associated
	 */
	get permissions() {
		return this.authenticationData.permissions;
	}

	/**
	 * Get the locations array associated to the session
	 *
	 * @return {Array<string>|undefined} The locations or undefined in case there are no locations associated
	 */
	get locations() {
		return this.authenticationData.locations;
	}

	/**
	 * Get if has Access To All locations associated to the session
	 *
	 * @return {boolean|undefined} The hasAccessToAllLocations field or undefined in case there are no hasAccessToAllLocations associated
	 */
	get hasAccessToAllLocations() {
		return this.authenticationData.hasAccessToAllLocations;
	}

	/**
	 * Get the warehousesIds array associated to the session
	 *
	 * @return {Array<string>|undefined} The warehousesIds or undefined in case there are no warehousesIds associated
	 */
	get warehousesIds() {
		return this.authenticationData.warehousesIds;
	}

	/**
	 * Fetch the client data from your own DB
	 *
	 * @return {Promise<import('./client').Client>} Resolves to the client object with the `getInstance()` method injected
	 * @throws {ApiSessionError} If an error occurs while fetching the client data
	 */
	get client() {

		if(this._client)
			return Promise.resolve(this.inject(this._client));

		if(!this.clientCode)
			return Promise.resolve();

		return ClientHandler
			.getByCode(this.clientCode, this.useClientCache)
			.then(client => client && this.inject(client));
	}

	/**
	 * Gets an instance injected with the session
	 *
	 * @template C
	 * @param {{new(): C}} TheClass The class
	 * @return {C} The instance.
	 */
	getSessionInstance(TheClass, ...params) {
		const instance = new TheClass(...params);
		instance.session = this;

		return instance;
	}

	/**
	 * Inject the client with methods to propagate this session
	 *
	 * @param {import('./client').Client} client The client data
	 * @param {import('./client').Client} The client data injected
	 */
	inject(client) {
		client.getInstance = this.getSessionInstance.bind(this);

		return client;
	}

	/**
	 * Validate if the session has access to the location
	 *
	 * @param {string} locationId Location Id
	 * @returns {Boolean}
	 */
	validateLocation(locationId) {
		return !!this.hasAccessToAllLocations || (!!locationId && Array.isArray(this.locations) && this.locations.includes(locationId));
	}

	/**
	 * Validate if the session has access to the warehouse
	 *
	 * @param {string} warehouseIds Warehouse Id
	 * @returns {Boolean}
	 */
	validateWarehouse(warehouseIds) {
		return !!this.hasAccessToAllLocations || (!!warehouseIds && Array.isArray(this.warehousesIds) && this.warehousesIds.includes(warehouseIds));
	}
};


lib/client.js
'use strict';

const logger = require('lllog')();

const path = require('path');

const ApiSessionError = require('./api-session-error');

/**
 * @typedef Client
 * @property {string | undefined} id
 * @property {string} code
 */

module.exports = class ClientHandler {

	static get clientsCache() {

		if(!this._clientsCache)
			this._clientsCache = {};

		return this._clientsCache;
	}

	/**
	 * @param {string} clientCode
	 * @param {boolean} useClientCache
	 * @return {Promise<Client>}
	 */
	static async getByCode(clientCode, useClientCache) {

		const cachedClient = useClientCache && this.getClientFromCache(clientCode);
		if(cachedClient) {
			logger.debug(`Client ${clientCode} fetched from cache`);
			return cachedClient;
		}

		const modelClient = this.getInstance();

		try {

			const clientsFound = await modelClient.getBy('code', clientCode, { limit: 1 });

			if(clientsFound.length) {
				const [client] = clientsFound;
				this.updateClientCache(client);
				return client;
			}

		} catch(e) {
			throw new ApiSessionError(e, ApiSessionError.codes.INTERNAL_ERROR);
		}

		throw new ApiSessionError(`Client not found for code ${clientCode}`, ApiSessionError.codes.CLIENT_NOT_FOUND);
	}

	/**
	 * Gets the client data from local cache if it exists.
	 *
	 * @return {object|undefined} The client data or undefined if it isn't cached
	 */
	static getClientFromCache(clientCode) {

		const cachedClient = this.clientsCache[clientCode];

		if(!cachedClient || (Date.now() >= cachedClient.expirationTime))
			return;

		return cachedClient.client;
	}

	/**
	 * Updates a client in the local cached
	 *
	 * @param {object} client The client data
	 * @return {object} The client received as argument
	 */
	static updateClientCache(client) {

		logger.debug(`Client ${client.code} saved in cache`);

		this.clientsCache[client.code] = {
			expirationTime: Date.now() + (10 * 60 * 1000), // 10 minutes cache
			client
		};
	}

	/**
     * Returns an instance model from the service.
     */
	static getInstance() {

		if(this._modelInstance)
			return this._modelInstance;

		const modelPath = this.getRelativePath();

		try {
			// eslint-disable-next-line global-require, import/no-dynamic-require
			const TheModelClass = require(modelPath);
			this._modelInstance = new TheModelClass();
			return this._modelInstance;
		} catch(e) {
			throw new ApiSessionError(`Invalid Model Client. Must be in ${modelPath}.`,
				ApiSessionError.codes.INVALID_MODEL_CLIENT);
		}
	}

	static getRelativePath() {
		return path.join(process.cwd(), process.env.MS_PATH || '', 'models', 'client');
	}
};
